File: pkg/service/subject_service.go
package service

import (
	"errors"
	"teaching-management/pkg/model"
	"teaching-management/pkg/repository"

	"github.com/google/uuid"
)

type SubjectService struct {
	subjectRepo *repository.SubjectRepository
}

func NewSubjectService(subjectRepo *repository.SubjectRepository) *SubjectService {
	return &SubjectService{subjectRepo: subjectRepo}
}

func (s *SubjectService) CreateSubject(userID uuid.UUID, name string) (*model.Subject, error) {
	if name == "" {
		return nil, errors.New("subject name cannot be empty")
	}

	subject := &model.Subject{
		UserID: userID,
		Name:   name,
	}

	err := s.subjectRepo.Create(subject)
	if err != nil {
		return nil, err
	}

	return subject, nil
}

func (s *SubjectService) GetSubjects(userID uuid.UUID) ([]model.Subject, error) {
	return s.subjectRepo.GetAll(userID)
}

func (s *SubjectService) UpdateSubject(userID uuid.UUID, id uuid.UUID, name string) error {
	if name == "" {
		return errors.New("subject name cannot be empty")
	}

	// Verify subject belongs to user
	subject, err := s.subjectRepo.GetByID(id)
	if err != nil {
		return err
	}
	if subject == nil {
		return errors.New("subject not found")
	}
	if subject.UserID != userID {
		return errors.New("unauthorized to update this subject")
	}

	subject.Name = name
	return s.subjectRepo.Update(subject)
}

func (s *SubjectService) DeleteSubject(userID uuid.UUID, id uuid.UUID) error {
	subject, err := s.subjectRepo.GetByID(id)
	if err != nil {
		return err
	}
	if subject == nil {
		return errors.New("subject not found")
	}
	if subject.UserID != userID {
		return errors.New("unauthorized to delete this subject")
	}

	return s.subjectRepo.Delete(id, userID)
}

File: src/services/subject.service.ts
import { SubjectRepository } from '../repositories/interfaces.js';
import { Subject } from '../types/index.js';

export class SubjectService {
  constructor(private subjectRepo: SubjectRepository) {}

  async createSubject(userId: string, name: string): Promise<Subject> {
    if (!name || name.trim() === '') {
      throw new Error('subject name cannot be empty');
    }
    return this.subjectRepo.create({ user_id: userId, name: name.trim() });
  }

  async getSubjects(userId: string): Promise<Subject[]> {
    return this.subjectRepo.getAll(userId);
  }

  async updateSubject(userId: string, id: string, name: string): Promise<void> {
    if (!name || name.trim() === '') {
      throw new Error('subject name cannot be empty');
    }

    const existing = await this.subjectRepo.getById(id);
    if (!existing) {
      throw new Error('subject not found');
    }
    if (existing.user_id !== userId) {
      throw new Error('unauthorized to update this subject');
    }

    await this.subjectRepo.update({ id, user_id: userId, name: name.trim() });
  }

  async deleteSubject(userId: string, id: string): Promise<void> {
    const existing = await this.subjectRepo.getById(id);
    if (!existing) {
      throw new Error('subject not found');
    }
    if (existing.user_id !== userId) {
      throw new Error('unauthorized to delete this subject');
    }

    await this.subjectRepo.delete(id, userId);
  }
}
